src/compiler/scanner.rs

```diff
@@ -128,6 +157,63 @@ impl Scanner {
         token
     }
 
+    fn token_value(&self) -> &str {
+        self.token_value.as_ref().unwrap()
+    }
+
+    fn set_token_value(&mut self, token_value: &str) {
+        self.token_value = Some(token_value.to_string());
+    }
+
+    fn is_digit(&self, ch: char) -> bool {
+        ch >= CharacterCodes::_0 && ch <= CharacterCodes::_9
+    }
+
+    fn scan_number_fragment(&mut self) -> String {
+        let start = self.pos();
+        let result = "".to_string();
+        loop {
+            let ch = self.text().chars().nth(self.pos());
+            let ch = match ch {
+                Some(ch) => ch,
```

Comment:
You could move the following if-statement into the match as a guard, which I think would be more clear. (Doing this off the top of my head)

```rust
match ch {
  Some(ch) if self.is_digit(ch) = > {
    self.set_pos(self.pos() + 1);
    continue;
  },
  None => break;
}

Reply:
Ya this `match` doesn't exist in the Typescript version because if we reach the end of the source text (while parsing a number), `.charCodeAt()` returns `NaN` which can be passed along to the other helpers which are expecting a `number`, whereas in Rust they're expecting `char`'s rather than `Option<char>`'s so I need to "unwrap" (and stop parsing the number if it's `None`). I actually like these patterns where you can unwrap eg an `Option` without having to nest the unwrapped logic by having the `None` branch abort the control flow (eg by `panic!()`'ing or here `break`'ing)